source/common/grpc/google_async_client_impl.h

```diff
@@ -0,0 +1,223 @@
+  // Convert from Google gRPC client std::multimap metadata to Envoy Http::HeaderMap.
+  void metadataTranslate(const std::multimap<grpc::string_ref, grpc::string_ref>& grpc_metadata,
+                         Http::HeaderMap& header_map);
+  // The threading model for the Google gRPC C++ library is not directly compatible with Envoy's
+  // siloed model. We resolve this by issuing non-blocking asynchronous operations on the silo
+  // thread, and then synchronously blocking on a completion queue, cq_, on a distinct thread. When
+  // cq_ events are delivered, we cross-post to the silo dispatcher to continue the operation.
+  //
+  // Currently, we have an independent thread for each stream that manages incoming events on cq_.
+  //
+  // TODO(htuch): Ideally we would have no more than one completion thread per silo. Having less
+  // than one per silo would be bad, as we would have potential contention, but having more than one
+  // per silo has unnecessary thread overhead, and in the current implementation we have thread
+  // creation/teardown overhead on every gRPC stream or unary RPC, which is sub-optimal. It is
+  // however, simple to grok for an initial implementation.
+  void completionThread();
+  // Write the first PendingMessage in the write queue if non-empty.
+  void writeQueued();
+  // Deliver notification and update stats when the connection closes.
+  void notifyRemoteClose(Status::GrpcStatus grpc_status, Http::HeaderMapPtr trailing_metadata,
+                         const std::string& message);
+  // Cleanup and schedule stream for deferred deletion.
+  void cleanup();
+
+  // Pending serialized message on write queue. Only one Operation::Write is in-flight at any
+  // point-in-time, so we queue pending writes here.
+  struct PendingMessage {
+    // We serialize the message to a grpc::ByteBuffer prior to queueing.
+    // TODO(htuch): We shouldn't need to do a string serialization & copy steps here (effectively,
+    // two copies!), this can be done more efficiently with SerializeToZeroCopyStream as done by
+    // TensorFlow
+    // https://github.com/tensorflow/tensorflow/blob/f9462e82ac3981d7a3b5bf392477a585fb6e6912/tensorflow/core/distributed_runtime/rpc/grpc_serialization_traits.h#L189
+    // at the cost of some additional implementation complexity.
+    // Also see
+    // https://github.com/grpc/grpc/blob/5e82dddc056bd488e0ba1ba0057247ab23e442d4/include/grpc%2B%2B/impl/codegen/proto_utils.h#L42
+    // which gives us what we want for zero copy, but relies on grpc::internal details; we can't get
+    // a grpc_byte_buffer from grpc::ByteBuffer to use this.
+    PendingMessage(const Protobuf::Message& request, bool end_stream)
+        : slice_(
+              [&request] {
+                const std::string buf = request.SerializeAsString();
+                return grpc_slice_from_copied_buffer(buf.c_str(), buf.size());
+              }(),
+              grpc::Slice::STEAL_REF),
```

Comment:
Rather than dipping into gRPC core surface APIs (which are technically not stable APIs), can I suggest that you use a pure gRPC C++ API for this:

``` cpp
slice_(request.SerializeAsString())
```

Because grpc::Slice does have a constructor version that takes a string as argument